api/grpcserver/v2alpha1/smeshing_identities.go

```diff
@@ -0,0 +1,153 @@
+	pb.RegisterSmeshingIdentitiesServiceServer(server, s)
+}
+
+func (s *SmeshingIdentitiesService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return pb.RegisterSmeshingIdentitiesServiceHandlerServer(context.Background(), mux, s)
+}
+
+// String returns the name of this service.
+func (s *SmeshingIdentitiesService) String() string {
+	return "SmeshingIdentitiesService"
+}
+
+func (s *SmeshingIdentitiesService) PoetServices(
+	context.Context,
+	*pb.PoetServicesRequest,
+) (*pb.PoetServicesResponse, error) {
+	states := s.states.IdentityStates()
+
+	pbIdentities := make(map[types.NodeID]*pb.PoetServicesResponse_Identity)
+	nodeIdsToRequest := make([]types.NodeID, 0)
+
+	for desc, state := range states {
+		pbIdentities[desc.NodeID()] = &pb.PoetServicesResponse_Identity{
+			SmesherId: desc.NodeID().Bytes(),
+			Status:    statusMap[state],
+		}
+
+		if state != types.IdentityStateWaitForPoetRoundEnd {
+			continue
+		}
+		nodeIdsToRequest = append(nodeIdsToRequest, desc.NodeID())
+	}
+
+	if len(nodeIdsToRequest) != 0 {
+		regInfos, err := s.collectPoetInfos(nodeIdsToRequest)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		for nodeId, poetInfos := range regInfos {
+			identity := pbIdentities[nodeId]
+			identity.PoetInfos = maps.Values(poetInfos)
+		}
+	}
+	return &pb.PoetServicesResponse{Identities: maps.Values(pbIdentities)}, nil
+}
+
+func (s *SmeshingIdentitiesService) collectPoetInfos(nodeIds []types.NodeID) (
+	map[types.NodeID]map[string]*pb.PoetServicesResponse_Identity_PoetInfo,
+	error,
+) {
+	dbRegs, err := nipost.PoetRegistrations(s.db, nodeIds...)
+	if err != nil {
+		return nil, err
+	}
+
+	identityRegInfos := make(map[types.NodeID]map[string]*pb.PoetServicesResponse_Identity_PoetInfo)
+	for _, reg := range dbRegs {
+		poetInfos, ok := identityRegInfos[reg.NodeId]
+		if !ok {
+			poetInfos = make(map[string]*pb.PoetServicesResponse_Identity_PoetInfo)
+		}
+
+		var (
+			status  pb.RegistrationStatus
+			warning string
+		)
+
+		_, ok = s.configuredPoetServices[reg.Address]
+		switch {
+		case !ok:
+			status = pb.RegistrationStatus_RESIDUAL_REG
+			warning = poetsMismatchWarning
+		case reg.RoundID == "":
+			status = pb.RegistrationStatus_FAILED_REG
+		default:
+			status = pb.RegistrationStatus_SUCCESS_REG
+		}
+
+		poetInfos[reg.Address] = &pb.PoetServicesResponse_Identity_PoetInfo{
+			Url:                reg.Address,
+			PoetRoundEnd:       timestamppb.New(reg.RoundEnd),
+			RegistrationStatus: status,
+			Warning:            warning,
+		}
+
+		identityRegInfos[reg.NodeId] = poetInfos
+	}
+
+	for id, poetInfos := range identityRegInfos {
+		for poetAddr := range s.configuredPoetServices {
+			if _, ok := poetInfos[poetAddr]; !ok {
+				identityRegInfos[id][poetAddr] = &pb.PoetServicesResponse_Identity_PoetInfo{
+					Url:                poetAddr,
+					RegistrationStatus: pb.RegistrationStatus_NO_REG,
+				}
+			}
+		}
+	}
+	return identityRegInfos, nil
+}
```

Comment:
Without the change to `nipost.PoetRegistrations` that allows to query multiple nodeIDs at once the code greatly simplifies by quite a bit:

```go
func (s *SmeshingIdentitiesService) PoetServices(
	context.Context,
	*pb.PoetServicesRequest,
) (*pb.PoetServicesResponse, error) {
	states := s.states.IdentityStates()

	pbIdentities := make(map[types.NodeID]*pb.PoetServicesResponse_Identity)
	for desc, state := range states {
		pbIdentities[desc.NodeID()] = &pb.PoetServicesResponse_Identity{
			SmesherId: desc.NodeID().Bytes(),
			Status:    statusMap[state],
		}

		if state != types.IdentityStateWaitForPoetRoundEnd {
			continue
		}

		regInfos, err := s.collectPoetInfos(desc.NodeID())
		if err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
		pbIdentities[desc.NodeID()].PoetInfos = maps.Values(regInfos)
	}
	return &pb.PoetServicesResponse{Identities: maps.Values(pbIdentities)}, nil
}

func (s *SmeshingIdentitiesService) collectPoetInfos(nodeId types.NodeID) (
	map[string]*pb.PoetServicesResponse_Identity_PoetInfo,
	error,
) {
	dbRegs, err := nipost.PoetRegistrations(s.db, nodeId)
	if err != nil {
		return nil, err
	}

	identityRegInfos := make(map[string]*pb.PoetServicesResponse_Identity_PoetInfo)
	for _, reg := range dbRegs {
		var (
			status  pb.RegistrationStatus
			warning string
		)

		_, ok := s.configuredPoetServices[reg.Address]
		switch {
		case !ok:
			status = pb.RegistrationStatus_RESIDUAL_REG
			warning = poetsMismatchWarning
		case reg.RoundID == "":
			status = pb.RegistrationStatus_FAILED_REG
		default:
			status = pb.RegistrationStatus_SUCCESS_REG
		}

		identityRegInfos[reg.Address] = &pb.PoetServicesResponse_Identity_PoetInfo{
			Url:                reg.Address,
			PoetRoundEnd:       timestamppb.New(reg.RoundEnd),
			RegistrationStatus: status,
			Warning:            warning,
		}
	}

	for poetAddr := range s.configuredPoetServices {
		if _, ok := identityRegInfos[poetAddr]; !ok {
			identityRegInfos[poetAddr] = &pb.PoetServicesResponse_Identity_PoetInfo{
				Url:                poetAddr,
				RegistrationStatus: pb.RegistrationStatus_NO_REG,
			}
		}
	}
	return identityRegInfos, nil
}
```

Even if a node manages 100 identities this should speed wise still be comparable. Ideally everything would happen within a single TX which wouldn't require much of a change to the suggested code above.

Reply:
To do it within a single Tx you would do something like:

```go
func (s *SmeshingIdentitiesService) PoetServices(
	ctx context.Context,
	_ *pb.PoetServicesRequest,
) (*pb.PoetServicesResponse, error) {
	pbIdentities := make(map[types.NodeID]*pb.PoetServicesResponse_Identity)
	err := s.db.WithTx(ctx, func(tx sql.Transaction) error {
		for desc, state := range s.states.IdentityStates() {
			pbIdentities[desc.NodeID()] = &pb.PoetServicesResponse_Identity{
				SmesherId: desc.NodeID().Bytes(),
				Status:    statusMap[state],
			}

			if state != types.IdentityStateWaitForPoetRoundEnd {
				continue
			}

			regInfos, err := s.collectPoetInfos(tx, desc.NodeID())
			if err != nil {
				return status.Error(codes.Internal, err.Error())
			}
			pbIdentities[desc.NodeID()].PoetInfos = maps.Values(regInfos)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return &pb.PoetServicesResponse{Identities: maps.Values(pbIdentities)}, nil
}

func (s *SmeshingIdentitiesService) collectPoetInfos(db sql.Executor, nodeId types.NodeID) (
	map[string]*pb.PoetServicesResponse_Identity_PoetInfo,
	error,
) {
	dbRegs, err := nipost.PoetRegistrations(db, nodeId)
	if err != nil {
		return nil, err
	}

	identityRegInfos := make(map[string]*pb.PoetServicesResponse_Identity_PoetInfo)
	for _, reg := range dbRegs {
		var (
			status  pb.RegistrationStatus
			warning string
		)

		_, ok := s.configuredPoetServices[reg.Address]
		switch {
		case !ok:
			status = pb.RegistrationStatus_RESIDUAL_REG
			warning = poetsMismatchWarning
		case reg.RoundID == "":
			status = pb.RegistrationStatus_FAILED_REG
		default:
			status = pb.RegistrationStatus_SUCCESS_REG
		}

		identityRegInfos[reg.Address] = &pb.PoetServicesResponse_Identity_PoetInfo{
			Url:                reg.Address,
			PoetRoundEnd:       timestamppb.New(reg.RoundEnd),
			RegistrationStatus: status,
			Warning:            warning,
		}
	}

	for poetAddr := range s.configuredPoetServices {
		if _, ok := identityRegInfos[poetAddr]; !ok {
			identityRegInfos[poetAddr] = &pb.PoetServicesResponse_Identity_PoetInfo{
				Url:                poetAddr,
				RegistrationStatus: pb.RegistrationStatus_NO_REG,
			}
		}
	}
	return identityRegInfos, nil
}
```

and you would need to change the type of `s.db` from `sql.Executor` to `sql.StateDatabase`.